Rebase ElectrumG to the latest upstream version

R-Santev opened the pull request:
## Motivation

ElectrumG was not recently updated and it's good to have the latest features and bug fixes from electrum. That's why I have rebased the wallet to the latest master of electrum.


## Implementation

The rebase process included multiple conflicts and therefore this branch is recommended to be merged by force push to master as it is rebased.

tolak committed:
Add BTG unit dict define

tolak committed:
Add BTG unit dict define

tolak committed:
Add BTG equihash algo impl

tolak committed:
Add BTG equihash algo impl

tolak committed:
Add BTG checkpoints and servers infomation

tolak committed:
Add BTG checkpoints and servers infomation

tolak committed:
Replace network definition with BTG

tolak committed:
Replace network definition with BTG

tolak committed:
Reimplement block serialize/deserialize functions

tolak committed:
Reimplement block serialize/deserialize functions

tolak committed:
Fix typo errors

tolak committed:
Fix typo errors

tolak committed:
Fix runtime error

tolak committed:
Fix runtime error

tolak committed:
Support BTG transaction signature

tolak committed:
Support BTG transaction signature

tolak committed:
Update testnet checkpoints

tolak committed:
Update testnet checkpoints

tolak committed:
1) Portting BTG blockchain class implementation
2) Add leveldb header storage implementation
3) Add new leveldb header storage unit test

tolak committed:
1) Portting BTG blockchain class implementation
2) Add leveldb header storage implementation
3) Add new leveldb header storage unit test

tolak committed:
Add unit test for header chunk save/read/delete

tolak committed:
Add unit test for header chunk save/read/delete

tolak committed:
Fix function definition typo issue

tolak committed:
Fix function definition typo issue

tolak committed:
Remove leveldb folder when delete header filesd

tolak committed:
Remove leveldb folder when delete header filesd

tolak committed:
Replace bitcoin chunk size 2016 with Bitcoin Gold

tolak committed:
Replace bitcoin chunk size 2016 with Bitcoin Gold

tolak committed:
Update BTG testnet checkpoints

tolak committed:
Update BTG testnet checkpoints

tolak committed:
Fix nonce convert issue within function deserialize_header()

tolak committed:
Fix nonce convert issue within function deserialize_header()

tolak committed:
Add new testnet checkpoint

tolak committed:
Add new testnet checkpoint

tolak committed:
Add new interface to save latest height to db

tolak committed:
Add new interface to save latest height to db

tolak committed:
Restore Blockchain from header database if this is not first time running

tolak committed:
Restore Blockchain from header database if this is not first time running

Commit:
BTG transaction signing support

Commit:
BTG transaction signing support

tolak committed:
Add missing functions

tolak committed:
Add missing functions

tolak committed:
Delete unnecessary loging information

tolak committed:
Delete unnecessary loging information

tolak committed:
1) Hardcode max checkpoint header into constants.py
2) Fix known target computing issues

tolak committed:
1) Hardcode max checkpoint header into constants.py
2) Fix known target computing issues

tolak committed:
Add logging infomation

tolak committed:
Add logging infomation

tolak committed:
Update checkpoints

tolak committed:
Update checkpoints

tolak committed:
Add available server url

tolak committed:
Add available server url

tolak committed:
Rewrite block synchronization and chunk verification

tolak committed:
Rewrite block synchronization and chunk verification

tolak committed:
Add testnet and mainnet checkpoints generation script

tolak committed:
Add testnet and mainnet checkpoints generation script

tolak committed:
Replace some string <Electrum> with <ElectrumG>

tolak committed:
Replace some string <Electrum> with <ElectrumG>

tolak committed:
Fix ElectrumG binary build issues

tolak committed:
Fix ElectrumG binary build issues

tolak committed:
Replace header storage with shelve

tolak committed:
Replace header storage with shelve

tolak committed:
Fix bytes encode issue

tolak committed:
Fix bytes encode issue

tolak committed:
Remove dependency pyblake2, replace with buildin module hashlib

tolak committed:
Remove dependency pyblake2, replace with buildin module hashlib

tolak committed:
Replace header storage database with diskcache (thread safe kv db)

tolak committed:
Replace header storage database with diskcache (thread safe kv db)

tolak committed:
Download diskcache during compiling on wine

tolak committed:
Download diskcache during compiling on wine

tolak committed:
Update header path config options

tolak committed:
Update header path config options

tolak committed:
Replace icons

tolak committed:
Replace icons

tolak committed:
Fix package base name on wine

tolak committed:
Fix package base name on wine

R-Santev committed:
Remove COPY aliyun from windows binaries Docker file

R-Santev committed:
Remove COPY aliyun from windows binaries Docker file

R-Santev committed:
Remove wrong diskcache and zbar installing

R-Santev committed:
Remove wrong diskcache and zbar installing

R-Santev committed:
Fix .exe and .ico names

R-Santev committed:
Fix .exe and .ico names

R-Santev committed:
Add dnspython to requirements.txt

R-Santev committed:
Add dnspython to requirements.txt

R-Santev committed:
Add diskcache as a dependency

R-Santev committed:
Add diskcache as a dependency

R-Santev committed:
Change BitcoinMainnet constants

R-Santev committed:
Change BitcoinMainnet constants

R-Santev committed:
Change the mainnet explorers

R-Santev committed:
Change the mainnet explorers

R-Santev committed:
Fixing file names and paths

R-Santev committed:
Fixing file names and paths

R-Santev committed:
Fix some bugs in building AppImage binaries

R-Santev committed:
Fix some bugs in building AppImage binaries

R-Santev committed:
Fix some namings

R-Santev committed:
Fix some namings

R-Santev committed:
Fix regKeys, namings and self._cleaned_up

R-Santev committed:
Fix regKeys, namings and self._cleaned_up

R-Santev committed:
Remove reg-keys which links the BTC url protocol

R-Santev committed:
Remove reg-keys which links the BTC url protocol

R-Santev committed:
Change testnet checkpoints

R-Santev committed:
Change testnet checkpoints

R-Santev committed:
Hide the 2FA option from wallet kinds

R-Santev committed:
Hide the 2FA option from wallet kinds

R-Santev committed:
Add the proper reg keys for btg

R-Santev committed:
Add the proper reg keys for btg

martin-key committed:
Adding the correct bip39 paths for BTG

martin-key committed:
Adding the correct bip39 paths for BTG

martin-key committed:
Fixing Trezor and Keepkey BTG name so it's recognized by the device.

martin-key committed:
Fixing Trezor and Keepkey BTG name so it's recognized by the device.

martin-key committed:
WIP Exchange rates

martin-key committed:
WIP Exchange rates

martin-key committed:
Fixing pull for price and adding bitcoingold URI for payment requests

martin-key committed:
Fixing pull for price and adding bitcoingold URI for payment requests

h4x3rotab commented:
Could you rebase to upstream/master?

```bash
git fetch upstream
git rebase upstream/master
git push --force
```

R-Santev committed:
Merge branch 'diskcache-latest' of https://github.com/R-Santev/electrum into diskcache-latest

h4x3rotab commented on contrib/build-wine/electrum.nsi:
```diff
@@ -6,9 +6,9 @@
-  !define PRODUCT_PUBLISHER "Electrum Technologies GmbH"
+  !define PRODUCT_NAME "ElectrumG"
+  !define PRODUCT_WEB_SITE "https://github.com/BTCGPU/electrum"
+  !define PRODUCT_PUBLISHER "BitcoinGold Team"
```
"The BTG Organization"

h4x3rotab commented on contrib/build-wine/electrum.nsi:
```diff
@@ -122,29 +122,29 @@ Section
+  WriteRegStr HKCU "Software\Classes\bitcoing" "" "URL:bitcoingold Protocol"
+  WriteRegStr HKCU "Software\Classes\bitcoing" "URL Protocol" ""
+  WriteRegStr HKCU "Software\Classes\bitcoing" "DefaultIcon" "$\"$INSTDIR\electrumg.ico, 0$\""
+  WriteRegStr HKCU "Software\Classes\bitcoing\shell\open\command" "" "$\"$INSTDIR\electrumg-${PRODUCT_VERSION}.exe$\" $\"%1$\""
```
Check btg core register path

h4x3rotab commented on contrib/build-wine/sources.list.aliyun:
```diff
@@ -0,0 +1,10 @@
+deb-src http://mirrors.aliyun.com/ubuntu/ bionic-security main restricted universe multiverse
+deb-src http://mirrors.aliyun.com/ubuntu/ bionic-updates main restricted universe multiverse
+deb-src http://mirrors.aliyun.com/ubuntu/ bionic-proposed main restricted universe multiverse
+deb-src http://mirrors.aliyun.com/ubuntu/ bionic-backports main restricted universe multiverse
```
Plz remove this file. It shouldn't be used

h4x3rotab commented on contrib/deterministic-build/requirements-hw.txt:
```diff
@@ -161,16 +161,17 @@ setuptools==49.6.0 \
+    --hash=sha256:88206b0eb87e6d677d424843ac5209e3fb9d0190d0ee169599165ec25e9d9115
+wheel==0.34.2 \
+    --hash=sha256:8788e9155fe14f54164c1b9eb0a319d98ef02c160725587ad60f14ddc57b6f96 \
+    --hash=sha256:df277cb51e61359aba502208d680f90c0493adec6f0e848af94948778aed386e
```
If no good reason, let's stick to the upstream version (they are also newer)

h4x3rotab commented on contrib/requirements/requirements.txt:
```diff
@@ -1,13 +1,11 @@
 qrcode
 protobuf>=3.12
+dnspython
```
Let's keep dnspython the same as the upstream (move to the bottom, and keep the comments)

h4x3rotab commented on electrum/base_wizard.py:
```diff
@@ -145,7 +145,7 @@ def new(self):
         wallet_kinds = [
             ('standard',  _("Standard wallet")),
-            ('2fa', _("Wallet with two-factor authentication")),
+            # ('2fa', _("Wallet with two-factor authentication")),
```
```suggestion
            # TODO: fix the 2fa
            # ('2fa', _("Wallet with two-factor authentication")),
```

h4x3rotab commented on electrum/blockchain.py:
```diff
@@ -246,7 +351,7 @@ def get_name(self) -> str:
         header_hash = hash_header(header)
         height = header.get('block_height')
-        return self.check_hash(height, header_hash)
+        return self.check_hash(height, header_hash)     
```
Remove trailing spaces

h4x3rotab commented on electrum/blockchain.py:
```diff
@@ -432,53 +543,50 @@ def assert_headers_file_available(self, path):
-        return self.read_header(height)
-
+        return self.read_header(height)   
+        
```
Spaces

h4x3rotab commented on electrum/constants.py:
```diff
@@ -39,136 +41,241 @@ def read_json(filename, default):
-GIT_REPO_URL = "https://github.com/spesmilo/electrum"
-GIT_REPO_ISSUES_URL = "https://github.com/spesmilo/electrum/issues"
+GIT_REPO_URL = "https://github.com/BTCGPU/electrum"
+GIT_ISSUE_URL = GIT_REPO_URL + "/issues"
```
```suggestion
GIT_REPO_ISSUES_URL = GIT_REPO_URL + "/issues"
```

And change the other files as well. This will be good for us to keep sync with the upstream

h4x3rotab commented on electrum/gui/kivy/main_window.py:
```diff
@@ -821,7 +821,9 @@ def init_ui(self):
         self.receive_screen = None
-        self.icon = os.path.dirname(KIVY_GUI_PATH) + "/icons/electrum.png"
+        self.requests_screen = None
+        self.address_screen = None
```
Are they still in use? If not, let's remove them

h4x3rotab commented on electrum/gui/qt/main_window.py:
```diff
@@ -815,11 +815,11 @@ def show_update_check(self, version=None):
             _("Please report any bugs as issues on github:<br/>"),
-            f'''<a href="{constants.GIT_REPO_ISSUES_URL}">{constants.GIT_REPO_ISSUES_URL}</a><br/><br/>''',
-            _("Before reporting a bug, upgrade to the most recent version of Electrum (latest release or git HEAD), and include the version number in your report."),
+            f'''<a href="{constants.GIT_ISSUE_URL}">{constants.GIT_ISSUE_URL}</a><br/><br/>''',
```
GIT_REPO_ISSUE_URL

h4x3rotab commented on electrum/lnaddr.py:
```diff
@@ -90,10 +90,15 @@ def bitarray_to_u5(barr):
+            wver = witness[0]
+            if wver > 16:
+                raise ValueError("Invalid witness version {}".format(witness[0]))
+            wprog = u5_to_bitarray(witness[1:])
```
Let's rollback to the upstream's implementation

h4x3rotab commented on electrum/transaction.py:
```diff
@@ -1865,7 +1870,7 @@ def serialize_preimage(self, txin_index: int, *,
             scriptCode = var_int(len(preimage_script) // 2) + preimage_script
             amount = int_to_hex(txin.value_sats(), 8)
-            nSequence = int_to_hex(txin.nsequence, 4)
+            nSequence = int_to_hex(0xfffffffd, 4)
```
I think this is incorrect. Let's roll back to the upstream version and check if tx singing works

R-Santev replied:
I changed it and it's working.

h4x3rotab commented on electrum/blockchain.py:
```diff
@@ -258,7 +363,7 @@ def check_hash(self, height: int, header_hash: str) -> bool:
             return False
 
-    def fork(parent, header: dict) -> 'Blockchain':
+    def fork(self, parent, header: dict) -> 'Blockchain':
```
```suggestion
    def fork(parent, header: dict) -> 'Blockchain':
```

R-Santev committed:
Apply the changes requested by H4

Change the PRODUCT_PUBLISHER name to "The BTG Organization"
Change one of the registry values in the electrum.nsi file
Remove the aliyun file
Stick the requirements-hw.txt file to the upstream version
Move the dnspython to the bottom of the requirements file and add the comments from the upstream
Add "TODO" for the 2FA in the base_wizard.py file
Remove trailing spaces in the blockchain.py file
Replace "GIT_ISSUE_URL" with "GIT_REPO_ISSUES_URL"
Remove "self.requests_screen = None" and "self.address_screen = None" because they are not used
Roll back a part of a function in the lnaddr.py file to the upstream
Roll back a variable value to the upstream in the transaction.py file
Remove "self" from the "fork" function arguments in the blockchain.py file

R-Santev committed:
Import with_lock in blockchain.py

h4x3rotab commented on electrum/lnaddr.py:
```diff
@@ -90,9 +90,9 @@ def bitarray_to_u5(barr):
-        wver, wprog_ints = segwit_addr.decode_segwit_address(currency, fallback)
-        if wver is not None:
+    if currency in [constants.BitcoinGoldMainnet.SEGWIT_HRP, constants.BitcoinGoldTestnet.SEGWIT_HRP]:
+        wver, wprog_ints = segwit_addr.decode_segwit_address(currency, fallback)	
+        if wver is not None:	
```
Trailing spaces

h4x3rotab commented on electrum/transaction.py:
```diff
@@ -1851,12 +1851,17 @@ def serialize_preimage(self, txin_index: int, *,
-        sighash = txin.sighash if txin.sighash is not None else SIGHASH_ALL
-        if sighash != SIGHASH_ALL:
-            raise Exception("only SIGHASH_ALL signing is supported!")
+        if txin.sighash is None:
+            txin.sighash = 0
+            # raise Exception("SigHash of txin couldn't be None!")
+        sighash = txin.sighash | SIGHASH_ALL | constants.net.SIGHASH_FORK_BTG
```
Let's switch back to the original impl, and always add `constants.net.SIGHASH_FORK_BTG` to the sighash flag

R-Santev committed:
Apply the changes requested by H4 (2)

Remove trailing spaces in the lnaddr.py file
Switch back to the original implementation and add  a constants.net.SIGHASH_FORK_BTG to the sighash flag in the transaction.py file
